packages/shared/src/util/index.ts

```diff
@@ -15,22 +15,68 @@ import { FeatureInterface, FeatureRule } from "back-end/types/feature";
 import { ExperimentReportVariation } from "back-end/types/report";
 import { VisualChange } from "back-end/types/visual-changeset";
 import { FeatureRevisionInterface } from "back-end/types/feature-revision";
+import { Environment } from "back-end/types/organization";
 
 export * from "./features";
 export * from "./saved-groups";
 
 export function getAffectedEnvsForExperiment({
   experiment,
+  orgEnvironments,
+  linkedFeatures,
 }: {
   experiment: ExperimentInterface | ExperimentInterfaceStringDates;
+  orgEnvironments: Environment[];
+  linkedFeatures?: FeatureInterface[];
 }): string[] {
+  if (!orgEnvironments.length) {
+    return [];
+  }
   // Visual changesets are not environment-scoped, so it affects all of them
-  if (experiment.hasVisualChangesets || experiment.hasURLRedirects)
+  // Also fallback to all envs if linkedFeatures is undefined, but the experiment does actually have linked features
+  if (
+    experiment.hasVisualChangesets ||
+    experiment.hasURLRedirects ||
+    (!linkedFeatures && !!experiment.linkedFeatures?.length)
+  )
     return ["__ALL__"];
 
-  // TODO: get actual environments for linked feature flags. We are being overly conservative here
-  if (experiment.linkedFeatures && experiment.linkedFeatures.length > 0) {
-    return ["__ALL__"];
+  if (linkedFeatures?.length) {
+    const envs = new Set<string>();
+    const orgEnvIds = orgEnvironments.map((e) => e.id);
+    linkedFeatures.forEach((linkedFeature) => {
+      const matches = getMatchingRules(
+        linkedFeature,
+        (rule) =>
+          (rule.type === "experiment-ref" &&
+            rule.enabled &&
+            rule.experimentId === experiment.id) ||
+          false,
+        orgEnvIds,
+        undefined,
+        // the boolean below skips environments if they are disabled on the feature
+        true
+      );
+      if (matches.length) {
+        const featureProject = linkedFeature.project || "";
+        matches.forEach((matchRule) => {
+          const environmentProjects =
```

Comment:
Have a look at `featureHasEnvironment()` (shared/src/util/features.ts) which might help simplify and standardize

Reply:
Oooh - that's a handy little utility function. Just pushed up some changes that now leverage this [here](https://github.com/growthbook/growthbook/pull/2835/commits/2bd90c7b28b433f389fb32a78ca9da0c96119a7c).